lib/ip_blocklist.js

```diff
@@ -0,0 +1,165 @@
+
+  function getBaseIp(addr, prefixLength) {
+    return ip.mask(addr, ip.fromPrefixLen(prefixLength))
+  }
+
+  function parseRow(ipColumn) {
+    // Ignore # lines
+    if (ipColumn.length > 0 && ipColumn[0] === '#') {
+      return
+    }
+
+    // Parse row for ip address components, supports 1.10.16.0/20 or 1.10.16.0.
+    // Converts them to a base IP address and a prefix length.
+    var tokens = ipColumn.split('/')
+    if (tokens.length === 2) {
+      var prefixLength = parseInt(tokens[1], 10)
+      return {
+        baseIp: getBaseIp(tokens[0], prefixLength),
+        prefixLength: prefixLength
+      }
+    }
+
+    return {
+      baseIp: tokens[0],
+      prefixLength: 32
+    }
+  }
+
+  function IPBlocklist() {
+    this.prefixLengths = []
+    this.ipsByPrefixLength = {}
+    this.fileLastModified = 0
+    this.pollInterval = null
+  }
+
+  IPBlocklist.prototype.load = function (filePath) {
+    var self = this
+
+    // Resolve file path to an absolute location
+    filePath = path.resolve(filePath)
+    self.filePath = filePath
+    self.fileName = path.basename(filePath)
+
+    var startTime = Date.now()
+    return statFile(filePath)
+      .then(function (fileStats) {
+        self.fileLastModified = fileStats.mtime
+        self.fileSize = fileStats.size
+        return readFile(filePath, 'utf8')
+      })
+      .then(function (data) {
+        return parse(data)
+      })
+      .then(function (rows) {
+        rows.forEach(function (row) {
+          var parsedData = parseRow(row[0])
+
+          if (parsedData) {
+            if (!self.ipsByPrefixLength[parsedData.prefixLength]) {
+              self.prefixLengths.push(parsedData.prefixLength)
+              self.ipsByPrefixLength[parsedData.prefixLength] = {}
+            }
+            self.ipsByPrefixLength[parsedData.prefixLength][parsedData.baseIp] = true
+          }
+        })
+
+        var endTime = Date.now()
+        log.info({
+          op: 'fxa.customs.blocklist.load',
+          fileName: self.fileName,
+          fileSize: self.fileSize,
+          loadedIn: (endTime - startTime)
+        })
+      })
+  }
+
+  IPBlocklist.prototype.clear = function () {
+    this.prefixLengths = []
+    this.ipsByPrefixLength = {}
+  }
+
+  IPBlocklist.prototype.contains = function (ipAddress) {
+    var self = this
+
+    var startTime = Date.now()
+    var endTime
+
+    log.info({
+      op: 'fxa.customs.blocklist.check',
+      ip: ipAddress,
+      fileName: self.fileName,
+      fileSize: self.fileSize
+    })
+
+    try {
+      return self.prefixLengths.some(function (prefixLength) {
+        var result = getBaseIp(ipAddress, prefixLength) in self.ipsByPrefixLength[prefixLength]
+
+        // If `hit` log metrics on time taken
+        if (result) {
+          endTime = Date.now()
+          log.info({op: 'fxa.customs.blocklist.hit', ip: ipAddress, fileName: self.fileName, fileSize: self.fileSize,
+            foundIn: (endTime - startTime)})
```

Comment:
nit: the indentation style here doesn't match our usual convention, which IIRC would expand this to have each item on its own line:

```
lof.info({
  op: ...,
  ip: ...,
  ...
})
```